usdGeom, usdShade: add validators related to UsdGeomSubsets

mattyjams opened the pull request:
The validators added in `usdGeom` and `usdShade` here check various aspects of `UsdGeomSubset`s:
- The set of all subset family names is fetched for an `Imageable`, and each family is checked for validity (using `UsdGeomSubset::ValidateFamily()`).
- `GeomSubset`s must be authored as direct descendants of an `Imageable` prim.
- If the `materialBind` subset family is authored on an `Imageable`, it is checked to ensure that it is of a restricted type (either `nonOverlapping` or `partition`), since it is invalid for an element of geometry to be bound to multiple materials. Those subsets are also checked to ensure that they are of element type `face`, since material bindings may
only be applied to geometric faces.
- If a `GeomSubset` has authored material bindings but no authored subset family name, it is suggested that the family name should be set to `materialBind` to ensure that the material bindings are visible to renderers. The material bindings will have no effect otherwise.

This replaces #2739, reworking what was there in the older compliance checker Python framework into the new C++ validation framework.

- [x] I have verified that all unit tests pass with the proposed changes
- [x] I have submitted a signed Contributor License Agreement

mattyjams committed:
usdShade: add "subsetMaterialBindFamilyName" validator token

mattyjams committed:
usdShade: implement and register the "SubsetMaterialBindFamilyName" validator

This validator checks whether a geom subset has authored material bindings
but no authored subset family name, in which case an error is emitted
indicating that the family name should be set to "materialBind" to ensure
that the material bindings are visible to renderers. The material bindings
will have no effect otherwise.

mattyjams committed:
usdShade: add a test for the "SubsetMaterialBindFamilyName" validator

mattyjams committed:
usdShade: add "subsetsMaterialBindFamily" validator token

mattyjams committed:
usdShade: implement and register the "SubsetsMaterialBindFamily" validator

This validator checks that if an Imageable prim has GeomSubsets of the
"materialBind" family, the family must be of a restricted type (either
"nonOverlapping" or "partition"), since it is invalid for an element of
geometry to be bound to multiple materials. Those subsets are also checked
to ensure that they are of element type "face", since material bindings may
only be applied to geometric faces.

mattyjams committed:
usdGeom: add validator tokens, including one for SubsetFamilies validator

We also include a "UsdGeomSubset" keyword token, as there will be a
collection of validators that check various aspects of geom subsets.

mattyjams committed:
usdGeom: implement and register the "SubsetFamilies" validator

This validator performs validation on all geom subset families authored
beneath an Imageable prim.

mattyjams committed:
usdGeom: add a test for the "SubsetFamilies" validator

mattyjams committed:
usdGeom: add "subsetParentIsImageable" validator token

mattyjams committed:
usdGeom: implement and register the "SubsetParentIsImageable" validator

This validator checks whether a geom subset prim is a direct descendant
of an Imageable prim.

mattyjams committed:
usdGeom: remove newline characters from subset validation failure reasons

This allows clients more flexibility when using the reason strings.

mattyjams committed:
usdGeom: add a test for the "SubsetParentIsImageable" validator

mattyjams committed:
usdShade: add a test for the "SubsetsMaterialBindFamily" validator

mattyjams commented on pxr/usd/usdGeom/subset.cpp:
```diff
@@ -749,7 +749,7 @@ UsdGeomSubset::ValidateSubsets(
                     valid = false;
                     if (reason) {
                         *reason += TfStringPrintf("Found overlapping index %d "
-                            "in GeomSubset at path <%s> at time %s.\n", index,
```
I think the newline characters here were a vestige from when we had planned to use this from the Python framework. We weren't consistent about including or not including the newline though, and it somewhat complicated things when trying to write the unit tests for the `usdGeom` subset families validator. I ended up just taking them all out so that clients can decide whether or not a trailing newline makes sense.

mattyjams commented on pxr/usd/usdGeom/validators.cpp:
```diff
@@ -0,0 +1,136 @@
+                /* elementType = */ TfToken(),
+                /* familyName = */ subsetFamilyName);
+
+        // Determine the element type of the family by looking at the
```
I wasn't completely certain about this. Is it possible/useful for the subsets of a given family to mix element types, or must they all have the same element type? I *think* the intention is the latter given what `ValidateFamily()` does here:
https://github.com/PixarAnimationStudios/OpenUSD/blob/5332d627cb28991e812acd67b8971e87367d9191/pxr/usd/usdGeom/subset.cpp#L831

But I don't actually think it's possible to end up in the code at the moment, since the `GetGeomSubsets()` call above it specifies an `elementType` which I don't think can be empty?

diyajoy replied:
The intention is for all the subsets in a family to have the same elementType!

The elementType in `GetGeomSubsets` can be empty, in which case all subsets in the family regardless of elementType will be returned. That being said, the `GetGeomSubsets()` call does pass in elementType, so great catch about that code never being hit! I'll update this code and the test for it separately.

mattyjams replied:
Great, thanks for clarifying @diyajoy!

Yeah, `GetGeomSubsets()` does allow for `elementType` to be empty, but the `_ValidateGeomType()` call at the top of `ValidateFamily()` basically ensures that it won't be, so our call to `GetGeomSubsets()` will be given a non-empty `elementType` and filter out all other element type subsets, preventing us from seeing whether there's a mix of element types.

Once we have that sorted out, I can add a test case for this in a follow-up PR to make sure we catch the case where a subset family contains subsets with differing element types and that we flag those as validation errors.

Thanks again!

diyajoy replied:
Edit: Please ignore, this comment was posted before your reply above!
Actually if it's easy enough, would you mind including a fix for the `ValidateFamily` issue in this PR? It should just be passing in an empty `TfToken` for the `elementType` in the call to `GetGeomSubsets()`!

diyajoy replied:
A follow-up PR sounds great, thank you!

Since `elementType` will always be non-empty, I'm thinking that if we pass in an empty `TfToken` directly instead of the `elementType`, we'll get all subsets in the family regardless of their type.

mattyjams replied:
It was indeed easy enough, so I just went ahead and made the change. :)

The fix is exactly as you described in d9cc6f344ee3c537cf597bc7dc5648f4777b183b, and then I incorporated a `mixedElementTypes` subset family in the unit test in c3d111878f93754775c483504f34e8a91cbdbeb7.

Thanks!

mattyjams commented on pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp:
```diff
@@ -23,40 +23,75 @@
     // UsdShadeValidators keyword.
-    UsdValidationRegistry &registry = UsdValidationRegistry::GetInstance();
-    UsdValidatorMetadataVector metadata = 
+    const std::set<TfToken> expectedUsdShadeValidatorNames = {
```
I ended up re-organizing this a little bit since we're interested in both all of the validators for the `UsdShadeValidators` keyword *and* the `UsdGeomSubset` keyword. The latter is a keyword we re-use from `usdGeom`'s validator keyword tokens.

jesschimein commented:
Filed as internal issue #USD-9764

jesschimein commented:
/AzurePipelines run

mattyjams commented:
Rebased on the current dev branch (c42f6d8a808156bc40734575f6db41b2acccf487) and resolved conflicts in `testUsdShadeValidators.cpp`.

jesschimein commented:
/AzurePipelines run

mattyjams committed:
usdGeom: query for subsets of all element types in UsdGeomSubset::ValidateFamily()

This ensures that we correctly detect the case where a subset family
consists of subsets that have differing element types, which should
result in a validation failure.

mattyjams committed:
usdGeom: adjust validation failure reason string for mixing subset element types

Referring to them as "GeomSubset" rather than just "Subset" is more
consistent with the other failure reasons, and both element type values
are now enclosed in single quotes.

jesschimein commented:
/AzurePipelines run

tallytalwar commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -192,13 +199,160 @@ _ShaderPropertyTypeConformance(const UsdPrim &usdPrim)
+            }
+        );
+    for (const UsdProperty& matBindingProperty : matBindingProperties) {
+        if (matBindingProperty.Is<UsdRelationship>()) {
```
Will be surprising to see a material:binding property on a prim which isn't a relationship. @mattyjams what do you think? I so we can clean this bit internally.

Thanks

tallytalwar replied:
Rather, we should have a core usdShade validator just for this ... check if a prim has a property starting with "material:binding" which isn't a relationship and if found, report a validation error.

And clean this check here.

mattyjams replied:
Good call, thanks @tallytalwar!

I added a `MaterialBindingRelationships` validator that issues errors about non-relationship properties in 77ef087ab883ff9264236ab860b4dacb78ef87df, and a test for it in 7556b56b9c772343191abf70b2267ef614357c12.

I think we need to keep the `if (matBindingProperty.Is<UsdRelationship>()) {` here though, because this subset validator is trying to determine if there are any valid material binding relationships. If there are any properties, this particular validator doesn't care about those and wants to ignore them, but the new validator will spit errors about that.

Does that sound ok to you?

tallytalwar replied:
Yeah that makes sense to me, thanks @mattyjams for jumping on the suggestion and implementing the new validator right away!

mattyjams committed:
usdShade: add "materialBindingRelationships" validator token

mattyjams committed:
usdShade: implement and register the "MaterialBindingRelationships" validator

This validator checks that all properties named "material:binding" or in the
"material:binding" namespace are relationships, as the expectation is that
they target Material prims.

mattyjams committed:
usdShade: add a test for the "MaterialBindingRelationships" validator

diyajoy commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -192,13 +243,164 @@ _ShaderPropertyTypeConformance(const UsdPrim &usdPrim)
+
+    // Check that all subsets belonging to the "materialBind" family are of
+    // element type "face", since material bindings may only be applied to
+    // geometric faces.
```
Since usdPhysics [also uses](https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usdPhysics/overview.dox#L403-L408) the materialBindingAPI, we now have use cases for binding materials to non-face subsets! Since [createMaterialBindSubsets](https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usdShade/materialBindingAPI.h#L909-L927) allows any elementType as well, I don't think we need this validation check anymore.

If you agree with that @mattyjams, I can go ahead and update your PR internally before pushing it!

mattyjams replied:
Oh, interesting! Sure, that sounds good to me, thanks @diyajoy!

jesschimein commented:
/AzurePipelines run
